src/pubsub/ua_pubsub_reader.c

```diff
@@ -1953,14 +1979,84 @@ decodeAndProcessNetworkMessage(UA_Server *server, UA_ReaderGroup *readerGroup,
                  (size_t)(connection->config->transportProfileUri.length)) == 0)) {
         paddingBytes = (MIN_PAYLOAD_SIZE_ETHERNET - ((*currentPosition) - previousPosition));
         (*currentPosition) += paddingBytes; /* During multiple receive, move the position to
-                             handle padding bytes */
+                                               handle padding bytes */
     }
 
 cleanup:
     UA_NetworkMessage_clear(&currentNetworkMessage);
     return rv;
 }
 
+static UA_StatusCode
+decodeHeadersAndDecryptNM(UA_Server *server, UA_ReaderGroup *readerGroup, UA_ByteString *buffer) {
```

Comment:
Do not put UA_NetworkMessage_decodeHeaders and verifyAndDecryptNetworkMessage into a shared function.
This forces users to decode the headers twice. Because this function does not return them.

Just call UA_NetworkMessage_decodeHeaders and verifyAndDecryptNetworkMessage separately in the parent context.